### ISSUE TITLE
Roll out KG2.8.1

### ISSUE THREAD

**Issue by acevedol** [RTXteam/RTX#1977]
_NOTE: To create a new issue based on this template, simply go to: https://github.com/RTXteam/RTX/issues/new?template=kg2rollout.md_

##### 1. Build and load KG2c:

- [x] merge `master` into the `kg2integration` branch
- [x] update the four hardcoded biolink version numbers in the `kg2integration` branch (as needed):
  - [x] in [code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml](../code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml)
  - [x] in [code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml](../code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml)
  - [x] in [code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.2_ARAX.yaml](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.2_ARAX.yaml)
  - [x] in [code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.2_KG2.yaml](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.2_KG2.yaml)
- [x] build a new KG2c on `buildkg2c.rtx.ai` from the `kg2integration` branch (how-to is [here](https://github.com/RTXteam/RTX/tree/master/code/kg2c#build-kg2canonicalized))
  - [x] before starting the build:
    - [x] make sure there is enough disk space available on `arax-databases.rtx.ai` (need at least 100G, ideally >120G). delete old KG2 database directories as needed (warn the team on Slack in advance).
    - [x] make sure to choose to build a new synonymizer in `kg2c_config.json`, as described in the how-to
  - [ ] after the build is done, verify it looks ok:
    - [ ] the 'build node' (node with ID `RTX:KG2c`) has the expected version number (KG2.X.Y)
    - [ ] `node_synonymizer.sqlite` should be around 15-20 GB
    - [ ] make sure `node_synonymizer.sqlite`'s last modified date is today (or whatever day the build was run)
    - [ ] make sure `kg2c_lite.json.gz`'s last modified date is today (or whatever day the build was run)
    - [x] the entire build runtime (synonymizer + KG2c) shouldn't have been more than 24 hours
    - [ ] the synonymizer and KG2c artifacts should have been auto-uploaded into the proper directory on `arax-databases.rtx.ai` (`/home/rtxconfig/KG2.X.Y`)
- [x] load the new KG2c into neo4j at http://kg2-X-Yc.rtx.ai:7474/browser/ (how to is [here](https://github.com/RTXteam/RTX/tree/master/code/kg2c#host-kg2canonicalized-in-neo4j))
  - [x] verify the correct KG2 version was uploaded by running this query: `match (n {id:"RTX:KG2c"}) return n`
  - [x] copy the auto-generated KG2c dump from the Neo4j-hosting instance to the database server like so:
    - `scp /home/ubuntu/kg2-build/kg2c.dump rtxconfig@arax-databases.rtx.ai:/home/rtxconfig/KG2.X.Y/extra_files`
  - [ ] update the KG2pre and KG2c Neo4j endpoints in `RTX/code/config_dbs.json` (push to `kg2integration` branch)
- [ ] upload the new `kg2c_lite_2.X.Y.json.gz` file to the [translator-lfs-artifacts](https://github.com/ncats/translator-lfs-artifacts/tree/main/files) repo
- [ ] load the new KG2c into Plover (available at http://kg2cplover.rtx.ai:9990)
  - [ ] update `config_dbs.json` to point to this new Plover (all maturity levels should point to it for now)
  

##### 2. Rebuild downstream databases:

The following databases should be rebuilt and copies of them should be put in `/home/rtxconfig/KG2.X.Y` on `arax-databases.rtx.ai`. Please use this kind of naming format: `mydatabase_v1.0_KG2.X.Y.sqlite`.

- [ ] NGD database
- [ ] refreshed DTD @chunyuma
- [ ] DTD model @chunyuma _(may be skipped - depends on the changes in this KG2 version)_
- [ ] DTD database @chunyuma _(may be skipped - depends on the changes in this KG2 version)_
- [ ] XDTD database

**NOTE**: As databases are rebuilt, `RTX/code/config_dbs.json` will need to be updated to point to their new paths! Push these changes to the `kg2integration` branch, unless the rollout of this KG2 version has already occurred, in which case you should push to `master` (but first follow the steps described [here](https://github.com/RTXteam/RTX/wiki/Config,-databases,-and-SFTP#config_dbsjson)). 

##### 3. Update the ARAX codebase:

All code changes should **go in the `kg2integration` branch**!

- [ ] regenerate the KG2c test triples file @acevedol
  - [ ] ensure the new KG2c Neo4j is currently running
  - [ ] check out the `kg2integration` branch and pull to get the latest changes (this is important for ensuring the correct KG2c Neo4j is used)
  - [ ] run [this script](https://github.com/RTXteam/RTX/blob/master/code/ARAX/KnowledgeSources/create_csv_of_kp_predicate_triples.py)
  - [ ] push the regenerated file to `RTX/code/ARAX/KnowledgeSources/RTX_KG2c_test_triples.json` (in the `kg2integration` branch)
- [ ] update Expand code as needed
- [ ] update any other modules as needed
- [ ] test everything together:
  - [ ] check out the `kg2integration` branch and pull to get the latest changes
  - [ ] locally set `force_local = True` in `ARAX_expander.py` (to avoid using the old KG2 API)
  - [ ] then run the entire ARAX pytest suite
  - [ ] address any failing tests
- [ ] update the KG2 and ARAX version numbers in the appropriate places (`openapi.yaml`, etc.) @edeutsch
  - [ ] Bump version for `RTX/code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml` in line 12 (`version:`); the major and minor release numbers are kept synchronous with the TRAPI version; just bump the patch release version (least significant digit)
  - [ ] Bump version for `RTX/code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml` in line 12 (`version:`); the first three digits are kept synchronous with the KG2 release version
  - [ ] Bump version number in `RTX/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.3_ARAX.yaml` on line 4 (`version:`); same as for the ARAX `openapi.yaml` file
  - [ ] Bump version number in `RTX/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.3_KG2.yaml` on line 4 (`version:`); same as for the KG2 `openapi.yaml` file
  

##### 4. Pre-upload databases:

Before rolling out, we need to pre-upload the new databases (referenced in `config_dbs.json`) to `arax.ncats.io` and the ITRB SFTP server. These steps can
be done well in advance of the rollout; it doesn't hurt anything to do them early.

- [ ] make sure `arax.ncats.io` has at least 100G of disk space free; delete old KG2 databases to free up space as needed (warn the team on slack first)
- [ ] `scp` the new databases from `arax-databases.rtx.ai` to `arax.ncats.io:/data/orangeboard/databases/KG2.X.Y`
- [ ] upload the new databases and their md5 checksums to ITRB's SFTP server using the steps detailed [here](https://github.com/RTXteam/RTX/wiki/Config,-databases,-and-SFTP#steps-for-all-databases-at-once)

##### 5. Do the rollout:

- [ ] merge `master` into `kg2integration`
- [ ] merge `kg2integration` into `master`
- [ ] roll `master` out to the various `arax.ncats.io` endpoints
- [ ] run the database manager
- [ ] run the pytest suite on the various endpoints
- [ ] verify each endpoint is running the new KG2 version by running the following JSON query and inspecting the returned node:
  - `{"nodes": {"n00": {"ids": ["RTX:KG2c"]}}, "edges": {}}`
- [ ] update our CI/CD testing instance with the new databases:
  - [ ] `ssh ubuntu@cicd.rtx.ai`
  - [ ] `cd RTX`
  - [ ] `git pull origin master`
  - [ ] `python3 code/ARAX/ARAXQuery/ARAX_database_manager.py --mnt --skip-if-exists --remove_unused`

##### 6. Final items/clean up:

- [ ] generate KGX files and upload them to the KGE Archive @acevedol
- [ ] turn off the old KG2c version's neo4j instance
- [ ] turn off the old KG2c version's plover instance
- [ ] turn off the new KG2pre version's neo4j instance
- [ ] upgrade the ITRB Plover endpoint (https://kg2cploverdb.ci.transltr.io) to this KG2 version and make the KG2 API start using it (instead of our self-hosted endpoint): 
    - [ ] update `kg_config.json` in the `main` branch of the Plover repo to point to the new `kg2c_lite_2.X.Y.json.gz` file (push this change)
    - [ ] wait about 45 minutes for the endpoint to rebuild and then run Plover tests to verify it's working
    - [ ] run the ARAX pytest suite with the NCATS endpoint plugged in (locally change the URL in `config_dbs.json` and set `force_local = True` in Expand)
    - [ ] if all tests pass, update `config_dbs.json` in `master` to point to the ITRB Plover endpoints (all maturity levels)
    - [ ] roll `master` out to the various endpoints on arax.ncats.io
    - [ ] turn off our plover endpoint and verify once more that ARAX is still working ok

**Comment by acevedol** [comment]
I am running into an issue with `ModuleNotFoundError: No module named 'textstat'`
I discussed with @amykglen and am working on installing `requirements.txt`. 
Currently, the installation exits with 
```
--- Logging error ---
Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/utils/logging.py", line 177, in emit
    self.console.print(renderable, overflow="ignore", crop=False, style=style)
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_vendor/rich/console.py", line 1673, in print
    extend(render(renderable, render_options))
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_vendor/rich/console.py", line 1305, in render
    for render_output in iter_render:
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/utils/logging.py", line 134, in __rich_console__
    for line in lines:
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_vendor/rich/segment.py", line 249, in split_lines
    for segment in segments:
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_vendor/rich/console.py", line 1283, in render
    renderable = rich_cast(renderable)
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_vendor/rich/protocol.py", line 36, in rich_cast
    renderable = cast_method()
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/self_outdated_check.py", line 130, in __rich__
    pip_cmd = get_best_invocation_for_this_pip()
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/utils/entrypoints.py", line 58, in get_best_invocation_for_this_pip
    if found_executable and os.path.samefile(
  File "/usr/lib/python3.9/genericpath.py", line 101, in samefile
    s2 = os.stat(f2)
FileNotFoundError: [Errno 2] No such file or directory: '/usr/bin/pip3.9'
Call stack:
  File "/home/ubuntu/.local/bin/pip", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/cli/main.py", line 70, in main
    return command.main(cmd_args)
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 101, in main
    return self._main(args)
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 223, in _main
    self.handle_pip_version_check(options)
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 190, in handle_pip_version_check
    pip_self_version_check(session, options)
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/self_outdated_check.py", line 236, in pip_self_version_check
    logger.warning("[present-rich] %s", upgrade_prompt)
  File "/usr/lib/python3.9/logging/__init__.py", line 1458, in warning
    self._log(WARNING, msg, args, **kwargs)
  File "/usr/lib/python3.9/logging/__init__.py", line 1589, in _log
    self.handle(record)
  File "/usr/lib/python3.9/logging/__init__.py", line 1599, in handle
    self.callHandlers(record)
  File "/usr/lib/python3.9/logging/__init__.py", line 1661, in callHandlers
    hdlr.handle(record)
  File "/usr/lib/python3.9/logging/__init__.py", line 952, in handle
    self.emit(record)
  File "/home/ubuntu/.local/lib/python3.9/site-packages/pip/_internal/utils/logging.py", line 179, in emit
    self.handleError(record)
Message: '[present-rich] %s'
Arguments: (UpgradePrompt(old='22.2.2', new='23.0.1'),)
```

**Comment by amykglen** [comment]
ah, nice - I don't think that was at play because we hadn't merged that commit into `kg2integration` yet

**Comment by amykglen** [comment]
ok @acevedol - I got a virtual environment working on `buildkg2c.rtx.ai` and got `requirements.txt` to install into it successfully.

to use it, first run this (from any directory on `buildkg2c.rtx.ai`):
```
pyenv activate rtx
```
then run scripts as usual, using the `python` command (and `pip`, if you ever need it).

so going forward always make sure to activate this virtual environment before running KG2c code. (we probably should have switched to a virtual environment long ago!)

we should probably add this step to the KG2c README...

**Comment by acevedol** [comment]
Thanks, @amykglen. That worked. 
I ran into a new issue, though. I updated `kg2c_config.json` with 
```
{
  "kg2pre_version": "2.8.1",
  "kg2pre_neo4j_endpoint": "kg2endpoint-kg2-8-1.rtx.ai",
  "biolink_version": "3.1.1",
  "upload_to_arax_databases.rtx.ai": true,
  "synonymizer": {
    "build": true,
    "synonymizer_db_version": "v1.0"
  },
  "kg2c": {
    "build": true,
    "kg2c_db_version": "v1.0",
    "use_nlp_to_choose_descriptions": true,
    "upload_to_s3": true,
    "start_from_kg2c_json": false,
    "use_local_kg2pre_tsvs": false
  }
}
```
So it looks like `synonymizer_db_version` replaced `name` and I'm not sure how to update it. I got the error: 
```
Traceback (most recent call last):
  File "/home/ubuntu/RTX/code/kg2c/build_kg2c.py", line 150, in <module>
    main()
  File "/home/ubuntu/RTX/code/kg2c/build_kg2c.py", line 100, in main
    raise ValueError(f"kg2c_config.json specifies that a new synonymizer should be built, but the "
ValueError: kg2c_config.json specifies that a new synonymizer should be built, but the synonymizer name provided already exists in /home/ubuntu/RTX/code/kg2c/../ARAX/NodeSynonymizer. You must enter a new synonymizer name.
```

**Comment by acevedol** [comment]
Okay, I found where to update the synonymizer name from above, but now there's a new error: 
```
2023-03-15 23:10:36,663 INFO: Building node synonymizer off of specified KG2..
+ set -e
+++ dirname /home/ubuntu/RTX/code/kg2c/build-synonymizer.sh
++ cd /home/ubuntu/RTX/code/kg2c
++ pwd
+ script_dir=/home/ubuntu/RTX/code/kg2c
+ code_dir=/home/ubuntu/RTX/code/kg2c/..
+ synonymizer_dir=/home/ubuntu/RTX/code/kg2c/../ARAX/NodeSynonymizer
+ kg2c_db_version=v1.0
+ kg2_version=2.8.1
+ cd /home/ubuntu/RTX/code/kg2c/../ARAX/NodeSynonymizer
+ rm -f sri_node_normalizer_requests_cache.sqlite
+ rm -f sri_node_normalizer_curie_cache.pickle
+ python3 -u dump_kg2_node_data.py
Traceback (most recent call last):
  File "/home/ubuntu/RTX/code/ARAX/NodeSynonymizer/dump_kg2_node_data.py", line 19, in <module>
    np.warnings.filterwarnings('ignore')
  File "/home/ubuntu/.pyenv/versions/rtx/lib/python3.9/site-packages/numpy/__init__.py", line 320, in __getattr__
    raise AttributeError("module {!r} has no attribute "
AttributeError: module 'numpy' has no attribute 'warnings'
Traceback (most recent call last):
  File "/home/ubuntu/RTX/code/kg2c/build_kg2c.py", line 150, in <module>
    main()
  File "/home/ubuntu/RTX/code/kg2c/build_kg2c.py", line 120, in main
    subprocess.check_call(["bash", "-x", f"{KG2C_DIR}/build-synonymizer.sh", kg2c_db_version, kg2_version])
  File "/home/ubuntu/.pyenv/versions/3.9.7/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['bash', '-x', '/home/ubuntu/RTX/code/kg2c/build-synonymizer.sh', 'v1.0', '2.8.1']' returned non-zero exit status 1.
```

**Comment by amykglen** [comment]
hmm - ok, I just pushed a change to `kg2integration` that I think fixes that most recent issue

**Comment by acevedol** [comment]
Thank you, @amykglen !

**Comment by amykglen** [comment]
re: the synonymizer-related problem - I'm not sure why it was mad... was there already a `node_synonymizer_KG2.8.1.sqlite` file in `code/ARAX/NodeSynonymizer`? that'd be pretty weird...

the way the naming of the synonymizer works now is it's derived based on the `kg2pre_version` and `synonymizer_db_version` specified in `kg2c_config.json` (thanks to Sundar's work). so if those are `2.8.1` and `v1.0`, respectively, then the synonymizer will be named `node_synonymizer_v1.0_KG2.8.1.sqlite`

**Comment by acevedol** [comment]
There's a little error in `build_kg2c.py` where the kg2 version in the node synonymizer name is hard coded: `    synonymizer_name = f"node_synonymizer_{synonymizer_db_version}_KG2.8.0.sqlite"`

**Comment by amykglen** [comment]
ohh, nice catch! did you already code up the fix or do you want me to push that?

**Comment by acevedol** [comment]
I left it because it looks like it's fixed in `master` (?) and the change might not have made it to integration yet. I need a sanity check though because I'm not sure how/when updates/merging works on this branch.

**Comment by amykglen** [comment]
ok cool - I just pushed the fix to `kg2integration`.

yeah, the branch situation is a little confusing - I merged `master` into `kg2integration` earlier today, so `kg2integration` pretty much has everything in `master` at this point (aside from anything committed to `master` in the last couple hours), but it also has some work that isn't yet in `master` (that Sundar did to improve the KG2c build, plus fixes I've pushed today). that stuff won't get merged into `master` until we roll out KG2.8.1.

**Comment by acevedol** [comment]
Got another error. 
```
WARNING: Beginning full NodeSynonymizer build process. This requires 54 GB of RAM. If you don't have 54 GB of RAM available, this would be a good time to stop the process!
INFO: Will use exception rename NCBIGENE:43 = AChE (human)
INFO: Will use exception rename UNIPROTKB:P22303 = AChE (human)
INFO: Will use exception rename RXNORM:1053143 = Stye(TM)
INFO: Will use exception rename ENSEMBL:ENSG00000087085 = AChE (human)
INFO: Reading SRI normalizer cache from sri_node_normalizer_curie_cache.pickle
ERROR returned with status 404 while retrieving supported types
<Response [404]>
Traceback (most recent call last):
  File "/home/ubuntu/RTX/code/kg2c/build_kg2c.py", line 150, in <module>
    main()
  File "/home/ubuntu/RTX/code/kg2c/build_kg2c.py", line 120, in main
    subprocess.check_call(["bash", "-x", f"{KG2C_DIR}/build-synonymizer.sh", kg2c_db_version, kg2_version])
  File "/home/ubuntu/.pyenv/versions/3.9.7/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['bash', '-x', '/home/ubuntu/RTX/code/kg2c/build-synonymizer.sh', 'v1.0', '2.8.1']' returned non-zero exit status 211.
```

**Comment by sundareswarpullela** [comment]
Hey @amykglen @acevedol , yeah its pretty much the same issue. I'll look and share the smart-api URL that worked last time.

**Comment by sundareswarpullela** [comment]
I couldn't find any endpoint where /get_semantic_types is working. I checked both https://nodenormalization-sri.renci.org/docs#/ and https://smart-api.info/ui/400f7c11028ff36f460af4ea85dc72f5#/default/get_semantic_types_handler_get_semantic_types_get, which were mentioned by Eric last time I faced this issue.

**Comment by edeutsch** [comment]
I reported the issue on the dev site. Maybe they will fix it

https://github.com/TranslatorSRI/NodeNormalization/issues/178

**Comment by edeutsch** [comment]
@amykglen is it possible/advisable to get this information via the BiolinkHelper module instead?

**Comment by acevedol** [comment]
I'm okay with trying that solution if no one has any objections.

**Comment by amykglen** [comment]
cool, I think it's worth a shot. I'm pretty sure they're not really used, looking at the code. 

ok @acevedol - I just commented out a couple lines in the `kg2integration` branch - want to try the build again?

**Comment by edeutsch** [comment]
Hi everyone, sorry to be late weighing in, I was off today. I have no objections. I do not recall why that information might have been needed at one point, but perhaps the switch to BiolinkHelper recently made it not necessary. If it's not used, fine to cut it out.

thanks!

**Comment by acevedol** [comment]
I ran into another issue:
```
2023-03-18 01:58:22,704 INFO: Downloading kg2-tsv-for-neo4j.tar.gz from the rtx-kg2 S3 bucket
download: s3://rtx-kg2/kg2-tsv-for-neo4j.tar.gz to RTX/code/kg2c/kg2-tsv-for-neo4j.tar.gz
2023-03-18 01:59:18,160 INFO: Unpacking kg2-tsv-for-neo4j.tar.gz..
tar (child): kg2-tsv-for-neo4j.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
Traceback (most recent call last):
  File "/home/ubuntu/RTX/code/kg2c/build_kg2c.py", line 150, in <module>
    main()
  File "/home/ubuntu/RTX/code/kg2c/build_kg2c.py", line 129, in main
    create_kg2c_files(args.test)
  File "/home/ubuntu/RTX/code/kg2c/create_kg2c_files.py", line 690, in create_kg2c_files
    subprocess.check_call(["tar", "-xvzf", kg2pre_tarball_name, "-C", local_tsv_dir_path])
  File "/home/ubuntu/.pyenv/versions/3.9.7/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['tar', '-xvzf', 'kg2-tsv-for-neo4j.tar.gz', '-C', '/home/ubuntu/RTX/code/kg2c/kg2pre_tsvs']' returned non-zero exit status 2.
```

I am not sure why the file cannot be opened. I see it in the directory. 
![Screen Shot 2023-03-18 at 3 37 45 PM](https://user-images.githubusercontent.com/59181061/226143781-918898a7-9bc3-4e0d-a600-59e7b2d0e9ba.png)

**Comment by acevedol** [comment]
Also note the node synonymizer for 2.8.1 is a bit smaller than the previous. 
![Screen Shot 2023-03-18 at 3 40 46 PM](https://user-images.githubusercontent.com/59181061/226143858-b339987f-36e5-46af-b0c9-b71c0ccbeba1.png)

**Comment by amykglen** [comment]
sorry, a little late here - so did you find a way around the `kg2-tsv-for-neo4j.tar.gz` error, @acevedol?

and thanks for the heads up about the synonymizer size. I think that's probably a small enough difference that we probably don't need to worry too much, but good to know...

**Comment by acevedol** [comment]
Yep @amykglen, I was just building it from the wrong directory, so I was the problem there. 

Good to know about the synonymizer size! Thank you for your feedback.

**Comment by acevedol** [comment]
Loaded KG2.8.1c onto neo4j at `http://kg2-8-1c.rtx.ai:7474/browser/` on `kg2canonicalized2.rtx.ai`
